### DISCUSSION TITLE
TypeScript error when building project with self-made package as project reference

### DISCUSSION THREAD

**Discussion by lukvermeulen** [vercel/next.js#36547] [Help]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

I have a monorepo setup with yarn workspaces and typescript project references consisting of 
- a nodejs server
- a schemas package (which works and compiles perfectly when included in the nodejs server)
- a NextJS 12 app

Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Pro
    Binaries:
      Node: 16.13.0
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 12.1.4
      react: 17.0.2
      react-dom: 17.0.2

### Describe the Bug

Running `yarn build` works perfectly and the linter also does not throw errors.

Running `yarn build` however results in throwing the following: 

```
yarn run v1.22.5
$ next build
info  - Loaded env from ..\packages\web-client\.env.local
Failed to compile.

../schemas/src/workspaces/index.ts:1:15
Type error: File '../packages/schemas/src/workspaces/workspaceSchema.ts' is not listed within the file list of project ''. Projects must list all files or use an 'include' pattern.

> 1 | export * from "./workspaceSchema";
    |               ^
  2 | 
error Command failed with exit code 1.
```

### Expected Behavior

I expect a normal working build working like building it from the nodejs project.

### To Reproduce

schemas tsconfig: 

```
{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "./dist",
    "rootDir": "./src",
    "composite": true
  },
  "include": ["src/**/*"]
}
```
nextjs tsconfig:
```
{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "target": "es5",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "incremental": true
  },
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx"],
  "exclude": ["node_modules"],
  "references": [{ "path": "../schemas" }]
}
```
the root tsconfig
```
{
  "compilerOptions": {
    "baseUrl": ".",
    "composite": true,
    "declaration": true,
    "module": "commonjs",
    "noImplicitAny": false,
    "removeComments": true,
    "noLib": false,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "target": "es6",

    "skipLibCheck": true,

    "sourceMap": true,
    "strict": true,
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,

    "paths": {
      "@/schemas": ["packages/schemas/src"],
      "@/schemas/*": ["packages/schemas/src/*"],
      "@/web-client/*": ["packages/schemas/src/*"]
    }
  },
  "exclude": ["node_modules", "**/*/lib", "**/*/dist"],
  "references": [
    { "path": "./packages/web-client" },
    { "path": "./packages/api-server" },
    { "path": "./packages/schemas" }
  ]
}
```

**Comment by igorpupkinable** [comment]
Looks like `next@10.2.3` does not work with composite projects.
Removing `composite` and `"references": [{ "path": "../../libs/xxx" }]` from `apps/my-nextjs-app/tsconfig.json` solved the issue.
This is not a good solution because now TS is unaware of dependencies for its `tsc --build` command.

**Reply by lukvermeulen** [reply]
I'll give that a shot with my project as well and see what happens, although I am using `next@12.1.x`. 
But I agree, this is not an ideal solution.

**Comment by Naveed-Iftikhar** [comment]
yes

Naveed
